rfc/002-fault-tolerance/README.md

```diff
@@ -0,0 +1,487 @@
+transport channel is out of scope.
+
+A transaction that writes one or more partitions needs to store additional
+metadata, containing the offsets this partition is based on. This atomic commit
+of both the transaction and its source position allows the source to pick up
+exactly at the point of failure, without risk of re-ingesting twice or missing
+any events. It thus provides exactly-once semantics for our write path.
+
+In technical terms, we can extend the CAF Streaming message type to include a
+handle to the source and source specific position, and then trigger an ACK once
+the data has been persisted. VAST already implements a similar mechanism today
+that acknowledges when events arrived at an active partition, which (almost)
+guarantees that they are queryable.
+
+When an exising source reconnects to an index component after a fault, it
+needs to request the most recently acknowlegded source position from the
+catalog and advance to the returned position to prevent double ingestion.
+
+### Read Path
+
+For historical queries, we can achieve fault-tolerance on the read part by
+moving query execution into a dedicated process. This requires that the catalog
+is the sole owner of partitions, and that the catalog returns the set of
+candidate partitions relevant to answer a query.
+
+Live queries bypass the catalog and would therefore need to implement ACK'ing
+independently, at the granularity of table slices. We do not cover this process
+in detail, as it constitutes a subset of the ACK'ing for the write path outlined
+above.
+
+### Metastore
+
+The metastore primarily stores the state of partition data. We restrict our
+analysis to this use case. From the perspective of the catalog, the metastore
+API should expose itself as an actor that encapsulates the implementation
+details. This actor can rely on VAST's plugin architecture to implement the
+metastore-specific integration.
+
+#### Request-Response API
+
+The catalog requires the following interface from the metastore:
+
+1. `get: [uuid] → [partition]`
+2. `add: [partition] → ACK`
+3. `delete: [uuid] → ACK`
+4. `replace: [uuid], [partition] → ACK`
+
+The semantics of the mentioned types are as follows:
+
+- `partition`: metadata about a partition, such as its URL and its schema.
+- `uuid`: a globally unique identifier for a partition.
+- `ACK`: a confirmation about an action.
+
+The semantics the function signatures are as follows:
+
+1. `get`: retrieves a list of `partition` objects corresponding one-to-one to
+   the requested list of partition UUIDs.
+
+2. `add`: registers a list of new partitions.
+
+3. `delete`: removes a list of partition from the metastore (doesn't delete the
+   underlying data)
+
+4. `replace`: performs an atomic deletion and addition of partition data, by
+   deleting the given UUIDs and adding new partition metadata.
+
+In theory, the functionality of `replace` is a superset of `add` and `delete`.
```

Comment:
```suggestion
In praxis, the functionality of `replace` is a superset of `add` and `delete`.
```

Reply:
TODO: Look up if this is the correct phrasing.

Reply:
I was not able to find any source that claims "in praxis" is proper usage. I'll use "in practice" for now until you are able to convince me otherwise :wink:.